blosc/shuffle.c

```diff
@@ -7,585 +7,252 @@
-     when it is re-defined. */
-  #if !defined(__SSE2__) && defined(_MSC_VER) && \
-      (defined(_M_X64) || (defined(_M_IX86) && _M_IX86_FP >= 2))
-    #define __SSE2__
-  #endif
-#else
-  #include <stdint.h>
-  #include <inttypes.h>
-#endif  /* _WIN32 */
-
-#ifdef HAVE_AVX2
-void __attribute__((visibility ("hidden")))
-shuffle2_AVX2(uint8_t* dest, const uint8_t* src, size_t size);
-void __attribute__((visibility ("hidden")))
-shuffle4_AVX2(uint8_t* dest, const uint8_t* src, size_t size);
-void __attribute__((visibility ("hidden")))
-shuffle8_AVX2(uint8_t* dest, const uint8_t* src, size_t size);
-void __attribute__((visibility ("hidden")))
-shuffle16_AVX2(uint8_t* dest, const uint8_t* src, size_t size);
-void __attribute__((visibility ("hidden")))
-unshuffle2_AVX2(uint8_t* dest, const uint8_t* src, size_t size);
-void __attribute__((visibility ("hidden")))
-unshuffle4_AVX2(uint8_t* dest, const uint8_t* src, size_t size);
-void __attribute__((visibility ("hidden")))
-unshuffle8_AVX2(uint8_t* dest, const uint8_t* src, size_t size);
-void __attribute__((visibility ("hidden")))
-unshuffle16_AVX2(uint8_t* dest, const uint8_t* src, size_t size);
+/* If C11 is supported use it's built-in atomic support for initialization. */
+#if __STDC_VERSION__ >= 201112L
+  #include <stdatomic.h>
 #endif
 
-static int
-have_avx2(void)
-{
-#if !defined(__clang__) && defined(__GNUC__) && defined(__GNUC_MINOR__) && \
-    __GNUC__ >= 5 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 8)
-  return __builtin_cpu_supports("avx2");
-#else
-  return 0;
-#endif
-}
-
-
-/* The non-SIMD versions of shuffle and unshuffle */
-
-/* Shuffle a block.  This can never fail. */
-void _shuffle(size_t bytesoftype, size_t blocksize,
-              const uint8_t* _src, uint8_t* _dest)
-{
-  size_t i, j, neblock, leftover;
-
-  /* Non-optimized shuffle */
-  neblock = blocksize / bytesoftype;  /* Number of elements in a block */
-  for (j = 0; j < bytesoftype; j++) {
-    for (i = 0; i < neblock; i++) {
-      _dest[j*neblock+i] = _src[i*bytesoftype+j];
-    }
-  }
-  leftover = blocksize % bytesoftype;
-  memcpy(_dest + neblock*bytesoftype, _src + neblock*bytesoftype, leftover);
-}
-
-/* Unshuffle a block.  This can never fail. */
-void _unshuffle(size_t bytesoftype, size_t blocksize,
-                       const uint8_t* _src, uint8_t* _dest)
-{
-  size_t i, j, neblock, leftover;
-
-  /* Non-optimized unshuffle */
-  neblock = blocksize / bytesoftype;  /* Number of elements in a block */
-  for (i = 0; i < neblock; i++) {
-    for (j = 0; j < bytesoftype; j++) {
-      _dest[i*bytesoftype+j] = _src[j*neblock+i];
-    }
-  }
-  leftover = blocksize % bytesoftype;
-  memcpy(_dest+neblock*bytesoftype, _src+neblock*bytesoftype, leftover);
-}
-
+/*  Include hardware-accelerated shuffle/unshuffle routines based on
+    the target architecture. Note that a target architecture may support
+    more than one type of acceleration!*/
+#if defined(__AVX2__)
```

Comment:
this must not be defined in the portable file as it implies  the compiler has been allowed to use any ISA up to avx2 for the file